timetomodel/speccing.py

```diff
@@ -179,6 +176,10 @@ def load_series(
                     )
                 )
 
+        # interpolate after the frequency is set (setting the frequency may have created additional nan values)
```

Comment:
Lines 143ff (the checks for empty and null values) should be after the interpolation, should they not? Because we interpolate those null/nan values if requested and thus correct that data.  Maybe it can simply be moved down? Not sure what the perfect place for that check is, actually.

Reply:
This may be more complicated than I first thought. First the easy part:

- The check on whether or not the `pd.Series` is empty should remain where it is I think, because the time window check assumes that the data is not empty.

Then a quick fix to the hard part:

- If no `interpolation_config` is given, then we should check for NaN values and fail early. Commit [506c3fc](https://github.com/SeitaBV/timetomodel/pull/16/commits/506c3fc6ed48e185bc8db70a7df25cbd415d61c3) implements that fix.

So what is the hard part?

- It's resampling. The resampling logic was nicely implemented, but only lightly tested for downsampling (from quarterhourly to hourly values). No tests had been written for upsampling. Taking the mean value is a good default resample method for downsampling, but for upsampling this actually introduces NaN values into the pd.Series. A good default resample method for upsampling is  padding (forward filling). So we could pick a better default resample method based on whether we're downsampling (`mean`) or upsampling (`pad`).
- We should check for and deal with NaN values separately based on their origin: either data is missing in the database and we may want to interpolate, or data is missing due to upsampling (I can't say for sure that downsampling may not lead to NaN values, too). We should then move the interpolation logic to **before** resampling and check for NaN values after resampling again.
- The `resampling_config` currently doesn't allow you to specify a method for each case separately, and forces the user to choose a resample method when defining the model, without knowing whether it will be used for downsampling or upsampling. I would recommend splitting that special "aggregation" argument in the resampling_config into two special arguments: "upsampling_method" and "downsampling_method".
- Upsampling has this quirky behaviour (at least for how we apply it to time series), where the last bin is narrowed rather than split. For example, the last one hour is bin is narrowed to one quarterhour bin instead of four quarterhour bins. I think I've dealt with this before in timely-beliefs.

Right now, I've not resolved these things, and concluded that resampling only functions well for downsampling. Currently, when upsampling, the interpolation_config allows you to deal with NaN values, regardless of their origin, because the interpolation still happens after resampling.

Reply:
Moved to separate issue #17.